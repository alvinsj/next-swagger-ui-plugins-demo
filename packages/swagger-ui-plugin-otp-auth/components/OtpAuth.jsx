import React from 'react'
import PropTypes from 'prop-types'

export default class OtpAuth extends React.Component {
  static propTypes = {
    name: PropTypes.string.isRequired,
    authorized: PropTypes.object,
    getComponent: PropTypes.func.isRequired,
    schema: PropTypes.object.isRequired,
    authSelectors: PropTypes.object.isRequired,
    errSelectors: PropTypes.object.isRequired,
  }

  constructor(props, context) {
    super(props, context)

    const { otpJwtAuthSelectors, errActions, otpJwtAuthActions } = props.getSystem()

    this.errActions = errActions
    this.otpJwtAuthActions = otpJwtAuthActions

    let { name, schema } = this.props

    let email = otpJwtAuthSelectors.email() || ""
    let otp = otpJwtAuthSelectors.otp()|| ""

    this.state = {
      name,
      schema,
      email,
      otp
    }
  }

  sendOtp = () => {
    let { name } = this.props
    let { otpJwtAuthActions, errActions } = this

    this.setState({ otp: ""})
    errActions.clear({ authId: name })
    otpJwtAuthActions.sendOtp(this.state)
  }

  authorize = () => {
    let { name } = this.props
    let { otpJwtAuthActions, errActions } = this

    errActions.clear({ authId: name })
    otpJwtAuthActions.authorizeOtpToken(this.state)
  }

  onChange =(e) => {
    let { target : { dataset : { name }, value } } = e
    let state = {
      [name]: value
    }

    this.setState(state)
  }

  logout =(e) => {
    e.preventDefault()
    let { name } = this.props
    let { otpJwtAuthActions, errActions } = this

    this.setState({ email: "", otp: ""})
    errActions.clear({ authId: name })
    otpJwtAuthActions.logout([ name ])
  }

  render() {
    let { getComponent, authSelectors, errSelectors, name, getSystem } = this.props
    const Input = getComponent("Input")
    const Row = getComponent("Row")
    const Col = getComponent("Col")
    const Button = getComponent("Button")
    const AuthError = getComponent("authError")
    const { otpJwtAuthSelectors } = getSystem()

    let authorizedAuth = authSelectors.authorized()
    let isAuthorized = !!authorizedAuth.get(name)
    let isOtpSent = !!otpJwtAuthSelectors.isOtpSent()

    let errors = errSelectors.allErrors().filter( err => err.get("authId") === name)
    let disabled = authorizedAuth.size > 0 && !isAuthorized

    return (
      <>
        <Row>
          <strong>OTP Login</strong>
        </Row>
        <Row>
          <label htmlFor="jwt_email">Email:</label>
          <Col>
            <Input id="jwt_email"
                   type="email"
                   value={this.state.email}
                   required="required"
                   data-name="email"
                   onChange={ this.onChange }
                   disabled={ isAuthorized || disabled } />
          </Col>
        </Row>
        <Row>
          <label htmlFor="jwt_otp">OTP:</label>
          <Col>
            <Input id="jwt_otp"
                   type="text"
                   value={this.state.otp}
                   required="required"
                   data-name="otp"
                   onChange={ this.onChange }
                   disabled={ isAuthorized || disabled } />
          </Col>
        </Row>
        {
          (isOtpSent || errors.valueSeq().size > 0) && 
            <div className="auth-msg-wrapper">
              {
                errors.valueSeq().map( (error, key) => {
                  return <AuthError error={ error }
                                    key={ key }/>
                } )
              }
              {
                isOtpSent ? <div style={{ backgroundColor: "#eeffee", color: "green" }}>
                              <span>OTP sent. Please check your email inbox.</span>
                            </div>
                          : null
              }
            </div>
        }
        <div className="auth-btn-wrapper">
        { isAuthorized ? <Button className="btn modal-btn auth authorize" onClick={ this.logout }>Logout</Button>
                       : <div>
                           <Button className="btn modal-btn auth send-otp" onClick={ this.sendOtp } disabled={ !this.state.email || disabled }>Send OTP</Button>
                           <Button className="btn modal-btn auth authorize" onClick={ this.authorize } disabled={ !this.state.email || !this.state.otp || disabled }>Authorize</Button>
                         </div>
        }
        </div>
      </>
    )
  }
}
