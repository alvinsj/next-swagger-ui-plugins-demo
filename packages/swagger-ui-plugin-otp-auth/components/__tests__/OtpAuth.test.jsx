import React from 'react'
import OtpAuth from '../OtpAuth'
import renderer from 'react-test-renderer'
import { Map } from 'immutable'

describe('<OtpAuth />', () => {
  const otpJwtAuthSelectors = { 
    email: () => 'test@example.com', 
    otp: () => 'example-otp',
    isOtpSent: () => false
  }
  const authSelectors = {
    authorized: () => ({ get: () => false })
  }
  const errSelectors = {
    allErrors: () => new Map()
  }

  const getComponent = (name) => 
    function GetComponent({children, ...props}) { 
      return <mock-element id={name} data-props={props}>{children}</mock-element> 
    }

  const system = {
    otpJwtAuthSelectors, 
    authSelectors,
    errSelectors,
    getComponent
  }

  it('renders form', () => {
    const tree = renderer
    .create(
      <OtpAuth 
        {...system}
        getSystem={() => system}
        schema={{}}
        name="OtpAuth"
      />)
    .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders otp sent message', () => {
    otpJwtAuthSelectors.isOtpSent = () => true
    const tree = renderer
    .create(
      <OtpAuth 
        {...system}
        getSystem={() => system}
        schema={{}}
        name="OtpAuth"
      />)
    .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders error messages', () => {
    otpJwtAuthSelectors.isOtpSent = () => false
    errSelectors.allErrors = () => new Map({'error': {get: () => 'OtpAuth'}})

    const tree = renderer
      .create(
        <OtpAuth 
          {...system}
          getSystem={() => system}
          schema={{}}
          name="OtpAuth"
        />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders logout', () => {
    otpJwtAuthSelectors.isOtpSent = () => false
    errSelectors.allErrors = () => new Map()
    authSelectors.authorized = () => ({ get: () => true })

    const tree = renderer
      .create(
        <OtpAuth 
          {...system}
          getSystem={() => system}
          schema={{}}
          name="OtpAuth"
        />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
