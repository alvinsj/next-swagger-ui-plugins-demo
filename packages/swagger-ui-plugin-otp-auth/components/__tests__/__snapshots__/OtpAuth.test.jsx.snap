// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OtpAuth /> renders error messages 1`] = `
[
  <mock-element
    data-props={{}}
    id="Row"
  >
    <strong>
      OTP Login
    </strong>
  </mock-element>,
  <mock-element
    data-props={{}}
    id="Row"
  >
    <label
      htmlFor="jwt_email"
    >
      Email:
    </label>
    <mock-element
      data-props={{}}
      id="Col"
    >
      <mock-element
        data-props={
          {
            "data-name": "email",
            "disabled": false,
            "id": "jwt_email",
            "onChange": [Function],
            "required": "required",
            "type": "email",
            "value": "test@example.com",
          }
        }
        id="Input"
      />
    </mock-element>
  </mock-element>,
  <mock-element
    data-props={{}}
    id="Row"
  >
    <label
      htmlFor="jwt_otp"
    >
      OTP:
    </label>
    <mock-element
      data-props={{}}
      id="Col"
    >
      <mock-element
        data-props={
          {
            "data-name": "otp",
            "disabled": false,
            "id": "jwt_otp",
            "onChange": [Function],
            "required": "required",
            "type": "text",
            "value": "example-otp",
          }
        }
        id="Input"
      />
    </mock-element>
  </mock-element>,
  <div
    className="auth-msg-wrapper"
  >
    <mock-element
      data-props={
        {
          "error": {
            "get": [Function],
          },
        }
      }
      id="authError"
    />
  </div>,
  <div
    className="auth-btn-wrapper"
  >
    <div>
      <mock-element
        data-props={
          {
            "className": "btn modal-btn auth send-otp",
            "disabled": false,
            "onClick": [Function],
          }
        }
        id="Button"
      >
        Send OTP
      </mock-element>
      <mock-element
        data-props={
          {
            "className": "btn modal-btn auth authorize",
            "disabled": false,
            "onClick": [Function],
          }
        }
        id="Button"
      >
        Authorize
      </mock-element>
    </div>
  </div>,
]
`;

exports[`<OtpAuth /> renders form 1`] = `
[
  <mock-element
    data-props={{}}
    id="Row"
  >
    <strong>
      OTP Login
    </strong>
  </mock-element>,
  <mock-element
    data-props={{}}
    id="Row"
  >
    <label
      htmlFor="jwt_email"
    >
      Email:
    </label>
    <mock-element
      data-props={{}}
      id="Col"
    >
      <mock-element
        data-props={
          {
            "data-name": "email",
            "disabled": false,
            "id": "jwt_email",
            "onChange": [Function],
            "required": "required",
            "type": "email",
            "value": "test@example.com",
          }
        }
        id="Input"
      />
    </mock-element>
  </mock-element>,
  <mock-element
    data-props={{}}
    id="Row"
  >
    <label
      htmlFor="jwt_otp"
    >
      OTP:
    </label>
    <mock-element
      data-props={{}}
      id="Col"
    >
      <mock-element
        data-props={
          {
            "data-name": "otp",
            "disabled": false,
            "id": "jwt_otp",
            "onChange": [Function],
            "required": "required",
            "type": "text",
            "value": "example-otp",
          }
        }
        id="Input"
      />
    </mock-element>
  </mock-element>,
  <div
    className="auth-btn-wrapper"
  >
    <div>
      <mock-element
        data-props={
          {
            "className": "btn modal-btn auth send-otp",
            "disabled": false,
            "onClick": [Function],
          }
        }
        id="Button"
      >
        Send OTP
      </mock-element>
      <mock-element
        data-props={
          {
            "className": "btn modal-btn auth authorize",
            "disabled": false,
            "onClick": [Function],
          }
        }
        id="Button"
      >
        Authorize
      </mock-element>
    </div>
  </div>,
]
`;

exports[`<OtpAuth /> renders logout 1`] = `
[
  <mock-element
    data-props={{}}
    id="Row"
  >
    <strong>
      OTP Login
    </strong>
  </mock-element>,
  <mock-element
    data-props={{}}
    id="Row"
  >
    <label
      htmlFor="jwt_email"
    >
      Email:
    </label>
    <mock-element
      data-props={{}}
      id="Col"
    >
      <mock-element
        data-props={
          {
            "data-name": "email",
            "disabled": true,
            "id": "jwt_email",
            "onChange": [Function],
            "required": "required",
            "type": "email",
            "value": "test@example.com",
          }
        }
        id="Input"
      />
    </mock-element>
  </mock-element>,
  <mock-element
    data-props={{}}
    id="Row"
  >
    <label
      htmlFor="jwt_otp"
    >
      OTP:
    </label>
    <mock-element
      data-props={{}}
      id="Col"
    >
      <mock-element
        data-props={
          {
            "data-name": "otp",
            "disabled": true,
            "id": "jwt_otp",
            "onChange": [Function],
            "required": "required",
            "type": "text",
            "value": "example-otp",
          }
        }
        id="Input"
      />
    </mock-element>
  </mock-element>,
  <div
    className="auth-btn-wrapper"
  >
    <mock-element
      data-props={
        {
          "className": "btn modal-btn auth authorize",
          "onClick": [Function],
        }
      }
      id="Button"
    >
      Logout
    </mock-element>
  </div>,
]
`;

exports[`<OtpAuth /> renders otp sent message 1`] = `
[
  <mock-element
    data-props={{}}
    id="Row"
  >
    <strong>
      OTP Login
    </strong>
  </mock-element>,
  <mock-element
    data-props={{}}
    id="Row"
  >
    <label
      htmlFor="jwt_email"
    >
      Email:
    </label>
    <mock-element
      data-props={{}}
      id="Col"
    >
      <mock-element
        data-props={
          {
            "data-name": "email",
            "disabled": false,
            "id": "jwt_email",
            "onChange": [Function],
            "required": "required",
            "type": "email",
            "value": "test@example.com",
          }
        }
        id="Input"
      />
    </mock-element>
  </mock-element>,
  <mock-element
    data-props={{}}
    id="Row"
  >
    <label
      htmlFor="jwt_otp"
    >
      OTP:
    </label>
    <mock-element
      data-props={{}}
      id="Col"
    >
      <mock-element
        data-props={
          {
            "data-name": "otp",
            "disabled": false,
            "id": "jwt_otp",
            "onChange": [Function],
            "required": "required",
            "type": "text",
            "value": "example-otp",
          }
        }
        id="Input"
      />
    </mock-element>
  </mock-element>,
  <div
    className="auth-msg-wrapper"
  >
    <div
      style={
        {
          "backgroundColor": "#eeffee",
          "color": "green",
        }
      }
    >
      <span>
        OTP sent. Please check your email inbox.
      </span>
    </div>
  </div>,
  <div
    className="auth-btn-wrapper"
  >
    <div>
      <mock-element
        data-props={
          {
            "className": "btn modal-btn auth send-otp",
            "disabled": false,
            "onClick": [Function],
          }
        }
        id="Button"
      >
        Send OTP
      </mock-element>
      <mock-element
        data-props={
          {
            "className": "btn modal-btn auth authorize",
            "disabled": false,
            "onClick": [Function],
          }
        }
        id="Button"
      >
        Authorize
      </mock-element>
    </div>
  </div>,
]
`;
